x-pack/plugins/task_manager/server/kibana_discovery_service/kibana_discovery_service.ts

```diff
@@ -0,0 +1,164 @@
+import { BACKGROUND_TASK_NODE_SO_NAME } from '../saved_objects';
+import { BackgroundTaskNode } from '../saved_objects/schemas/background_task_node';
+
+interface DiscoveryServiceParams {
+  currentNode: string;
+  savedObjectsRepository: ISavedObjectsRepository;
+  logger: Logger;
+}
+
+interface DiscoveryServiceUpsertParams {
+  id: string;
+  lastSeen: string;
+}
+
+export const discoveryInterval = 1000 * 10;
+export const cleanupInterval = 1000 * 60;
+export const cleanupLookBack = '5m';
+export const getActiveNodesLookBack = '30s';
+
+export class KibanaDiscoveryService {
+  private currentNode: string;
+  private discoveryStarted = false;
+  private cleanupStarted = false;
+  private savedObjectsRepository: ISavedObjectsRepository;
+  private logger: Logger;
+
+  constructor({ currentNode, savedObjectsRepository, logger }: DiscoveryServiceParams) {
+    this.savedObjectsRepository = savedObjectsRepository;
+    this.logger = logger;
+    this.currentNode = currentNode;
+  }
+
+  private async upsertCurrentNode({ id, lastSeen }: DiscoveryServiceUpsertParams) {
+    await this.savedObjectsRepository.update<BackgroundTaskNode>(
+      BACKGROUND_TASK_NODE_SO_NAME,
+      id,
+      {
+        id,
+        last_seen: lastSeen,
+      },
+      { upsert: { id, last_seen: lastSeen }, refresh: false }
+    );
+  }
+
+  private async deleteInactiveNodes() {
+    const { saved_objects: inactiveNodes } =
+      await this.savedObjectsRepository.find<BackgroundTaskNode>({
+        type: BACKGROUND_TASK_NODE_SO_NAME,
+        perPage: 10000,
+        page: 1,
+        filter: `${BACKGROUND_TASK_NODE_SO_NAME}.attributes.last_seen < now-${cleanupLookBack}`,
+      });
+
+    if (inactiveNodes.length > 0) {
+      const nodesToDelete = inactiveNodes.map((node) => ({
+        type: BACKGROUND_TASK_NODE_SO_NAME,
+        id: node.attributes.id,
+      }));
+      await this.savedObjectsRepository.bulkDelete(nodesToDelete, { force: true, refresh: false });
+      this.logger.info(
+        `Inactive Kibana nodes: ${nodesToDelete.map(
+          (node) => node.id
+        )}, have been successfully deleted`
+      );
+    }
+  }
+
+  private async scheduleDeleteInactiveNodes() {
+    try {
+      await this.deleteInactiveNodes();
+      if (!this.cleanupStarted) {
+        this.logger.info('Kibana Discovery Service - Cleanup - has been started');
+        this.cleanupStarted = true;
+      }
+    } catch (e) {
+      if (!this.cleanupStarted) {
+        this.logger.error(
+          `Kibana Discovery Service - Cleanup - couldn't be started and will be retried in ${cleanupInterval}ms. Error: ${e.message}`
+        );
+      } else {
+        this.logger.error(`Deleting inactive nodes failed. Error: ${e.message} `);
+      }
+    } finally {
+      setTimeout(async () => await this.scheduleDeleteInactiveNodes(), cleanupInterval);
+    }
+  }
+
+  private async scheduleUpsertCurrentNode() {
+    const lastSeenDate = new Date();
```

Comment:
Not sure if you've seen @rudolf 's [earlier thoughts on this](https://github.com/elastic/kibana/issues/93029#issuecomment-916352193). I think using a monotonically increasing number ([`process.hrtime.bigint()`](https://nodejs.org/api/process.html#processhrtimebigint)) may prove a bit more robust. I know this algo uses individual docs rather than reading from a single doc (so we avoid possible conflicts), but I still think this would be more robust.

Let me know what you think!

CC @rudolf

Reply:
Using `hrtime`, would we have to load all the documents (and wait the interval) before we know which Kibana nodes are running? Would the same pattern apply to determine which Kibana nodes are not running and can be cleaned up?

Reply:
Yes, it moves queries like "who is currently active" to Kibana and requires tracking the timing diffs for each node.

I'm anticipating two things (1) weirdness with time reported from Node.js (2) we'll probably want to hold information about all nodes in Kibana memory anyway.

But I definitely don't want to raise a blocker here or jump the gun, I've not spent a lot of time thinking about this problem and just wanted get your thoughts 😄

Reply:
I can see using `hrtime` as a better option. I'd want someone to learn more about the unknowns before using it in production. Perhaps we can implement this in the future by adding a new field? I'm not too concerned about scenarios where the clocks would be out of sync; our task management system relies on clocks being synced already, and we haven't seen this issue come up yet (we'll have bigger issues to fix with our task management to support that 🙈).

Reply:
I created https://github.com/elastic/kibana/issues/188465 to track moving to process.hrtime.